### ISSUE TITLE
TraceLogger is not properly cleaning up TelemetryConsumers during an Uninitialize

### ISSUE THREAD

**Issue by bsmatt** [dotnet/orleans#1471]
This particularly a problem for FileTelemetryConsumer as the StreamWriter is not closed properly leaving a lock on the file.

I attempted to properly close the StreamWriter but that is causing a shutdown exception because stuff is attempting to use the TraceLogger after it has been uninitialized

**Comment by gabikliot** [comment]
This is really messed up!

Apparently, with introduction of TelemetryConsumers, we now have duplicate code:
1) We have `ILogConsumer`, which do nothing now, since we have zero implementations of this interface. The log event is delivered to them, they are closed and flushed correctly, but alas - since none of the actual backends is implemented as  `ILogConsumer`, this code is essentially dead and useless.
2) `ITraceTelemetryConsumer`. We have file and console ITraceTelemetryConsumers, and all the logging is done via them, BUT they are not closeable and not flushable. Plus their Log interface is one string and not an interface like of `ILogConsumer` with all Log params apart, which is a rich semantical interface.Not to mention the code duplication between the two things.

Definitely a regression compared to the situation before TelemetryConsumers.

**Comment by sergeybykov** [comment]
`ILogConsumer` was left for backward compatibility until we flash out the telemetry story, by design.

>Plus their Log interface is one string and not an interface like of  ILogConsumer  with all Log params apart, which is a rich semantical interface.

`ITraceTelemetryConsumer` has overloads that take a dictionary of properties, which is more flexible than `ILogConsumer.Log`.

**Comment by bsmatt** [comment]
@galvesribeiro wanted to be tagged on this issue.

**Comment by bsmatt** [comment]
fwiw... I have a fix locally but I am not sure it is the best option. Basically I created a new ICloseableTelemetryConsumer interface and use that when the TraceLogger is shutting down. The problem with this is that as the silo or client is shutting down components are still attempting to write to the tracelogger. Since the FileTelemetryConsumer wasn't being cleaned up those messages where still making it to the logs. With my fix, as soon as the TraceLogger is shut down all logs stop. It seems like shutting down the tracerlogger should happen later.

**Comment by bsmatt** [comment]
Here is my change....
https://github.com/bsmatt/orleans/commit/39a8332d28e2d779000dee02832c6b7da34d34f6

**Comment by sergeybykov** [comment]
I was thinking if we should simply add `Close()` and `Flush()` to `ITelemetryConsumer`. IClosable/IFlushable seem overkill to me. A non-closable/non-flushable plugin can simply do nothing in the methods.

**Comment by sergeybykov** [comment]
In case it wasn't already clear. I DO NOT think this is "messed up" at all. Just a minor missing facet in the new implementation that was not caught by our tests.

The TelemetryConsumers feature is a major step forward that unifies our logging and telemetry story, and makes it more easily connectable to a range of existing monitoring and telemetry facilities out there. Gutemberg who worked relentlessly on this feature while integrating lots of our feedback did an outstanding job of rationalizing the design and putting together the implementation that we all agreed with. He rightfully deserves lots of kudos from us for do that.

**Comment by gabikliot** [comment]
Of course the TelemetryConsumers  are a good feature and we are thankful to Gutemberg  for that.

The mess, in my view, is:
a) code duplication and two competing  implementations: TelemetryConsumers  and LogConsumer. If we already moved all logic from LogConsumer to TelemetryConsumers  (which we did), I don't see why not to clean it up.
b) TelemetryConsumers   CAN receive semantic data (good design), but as of now it receives all the semantic data that LogConsumer  used to receive before as one string. This is a regression in this particular aspect.

Basically, what I am saying is that "someone" (the core team in my view) should have continued the great work that Gutemberg started and finish that up, and not leave it "almost there".
